contrib/epee/include/net/abstract_tcp_server2.inl

```diff
@@ -81,860 +81,1049 @@ namespace net_utils
+        state.timers.general.cancel_expire = false;
+        if (state.timers.general.reset_expire) {
+          state.timers.general.reset_expire = false;
+          async_wait_timer();
+        }
+        else if (state.status == status_t::INTERRUPTED)
+          on_interrupted();
+        else if (state.status == status_t::TERMINATING)
+          on_terminating();
+      }
+      else if (state.status == status_t::RUNNING)
+        interrupt();
+      else if (state.status == status_t::INTERRUPTED)
+        terminate();
+    });
   }
-  //---------------------------------------------------------------------------------
-  template<class t_protocol_handler>
-  bool connection<t_protocol_handler>::add_ref()
-  {
-    TRY_ENTRY();
 
-    // Use safe_shared_from_this, because of this is public method and it can be called on the object being deleted
-    auto self = safe_shared_from_this();
-    if(!self)
-      return false;
-    //_dbg3("[sock " << socket().native_handle() << "] add_ref, m_peer_number=" << mI->m_peer_number);
-    CRITICAL_REGION_LOCAL(self->m_self_refs_lock);
-    //_dbg3("[sock " << socket().native_handle() << "] add_ref 2, m_peer_number=" << mI->m_peer_number);
-    ++m_reference_count;
-    m_self_ref = std::move(self);
-    return true;
-    CATCH_ENTRY_L0("connection<t_protocol_handler>::add_ref()", false);
-  }
-  //---------------------------------------------------------------------------------
-  template<class t_protocol_handler>
-  bool connection<t_protocol_handler>::release()
-  {
-    TRY_ENTRY();
-    boost::shared_ptr<connection<t_protocol_handler> >  back_connection_copy;
-    LOG_TRACE_CC(context, "[sock " << socket().native_handle() << "] release");
-    CRITICAL_REGION_BEGIN(m_self_refs_lock);
-    CHECK_AND_ASSERT_MES(m_reference_count, false, "[sock " << socket().native_handle() << "] m_reference_count already at 0 at connection<t_protocol_handler>::release() call");
-    // is this the last reference?
-    if (--m_reference_count == 0) {
-        // move the held reference to a local variable, keeping the object alive until the function terminates
-        std::swap(back_connection_copy, m_self_ref);
-    }
-    CRITICAL_REGION_END();
-    return true;
-    CATCH_ENTRY_L0("connection<t_protocol_handler>::release()", false);
-  }
-  //---------------------------------------------------------------------------------
-  template<class t_protocol_handler>
-  void connection<t_protocol_handler>::call_back_starter()
+  template<typename T>
+  void connection<T>::cancel_timer()
   {
-    TRY_ENTRY();
-    _dbg2("[" << print_connection_context_short(context) << "] fired_callback");
-    m_protocol_handler.handle_qued_callback();
-    CATCH_ENTRY_L0("connection<t_protocol_handler>::call_back_starter()", void());
+    if (not state.timers.general.wait_expire)
+      return;
+    state.timers.general.cancel_expire = true;
+    state.timers.general.reset_expire = false;
+    ec_t ec;
+    timers.general.cancel(ec);
```

Comment:
`cancel` overload with error code is deprecated. `cancel` overload without error code out parameter throws `system_error`